### PR TITLE
cql_metrics: Add metrics for CQL errors

### LINKED ISSUE #5859: add metrics for CQL errors
I think we could use a metric for when scylla replies with an error for a CQL request

right now we have various storage proxy timeouts, etc. but no (single) metric that would clearly say - Scylla replied with X amount of errors or fails per second that client had to handle

### PR DESCRIPTION
**Pull request by StarostaGit**
This change adds tracking of all the CQL errors that can be
raised in response to a CQL message from a client, as described
in the CQL v4 protocol and with Scylla's CDC_WRITE_FAILUREs
included.

Fixes #5859

### CODE REVIEW

File: transport/server.hh [hunk 1]
```diff
@@ -136,6 +136,9 @@ private:
         uint64_t execute_requests;
         uint64_t batch_requests;
         uint64_t register_requests;
+
+        // errors
```
Comment (dekimir) [inline transport/server.hh:140]: This comment is pointless; it contains no information beyond the code.

**Review by dekimir — APPROVED** [review]
LGTM, one nitpick.

File: transport/server.cc [hunk 2]
```diff
@@ -207,11 +207,68 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
         sm::make_derive("requests_shed", _stats.requests_shed,
                         sm::description("Holds an incrementing counter with the requests that were shed due to overload (threshold configured via max_concurrent_requests_per_shard). "
                                             "The first derivative of this value shows how often we shed requests due to overload in the \"CQL transport\" component.")),
-       sm::make_gauge("requests_memory_available", [this] { return _memory_available.current(); },
+        sm::make_gauge("requests_memory_available", [this] { return _memory_available.current(); },
                         sm::description(
                             seastar::format("Holds the amount of available memory for admitting new requests (max is {}B)."
                                             "Zero value indicates that our bottleneck is memory and more specifically - the memory quota allocated for the \"CQL transport\" component.", _max_request_size))),
+        
+        sm::make_derive("server_errors", _stats.errors[exceptions::exception_code::SERVER_ERROR],
```
Comment (nyh) [inline transport/server.cc:215]: I hope that someone who's an expert in the overall monitoring "experience" (@amnonh ?) can review the naming of these counters. Something that bothers me is that you're adding about 20 new metrics which have a common theme (types of CQL errors) and yet there is no common naming that could make that clear. Maybe we should name all of them "cql_error_*"? Or maybe they should all be one metric with different labels - something which will also allow the user to request the sum of all errors?
I'm just raising a question here. I don't know the answer.

File: transport/server.cc [hunk 3]
```diff
@@ -207,11 +207,68 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
...
+                        sm::description("Counts the total number of raised CQL SERVER errors.")),
```
Comment (nyh) [inline transport/server.cc:216]: nitpick: I don't know if the word "raised" is the right term. It's errors "returned" to the client, no? Although I guess "raised" will be understood as well, so maybe it's ok.

File: transport/server.cc [hunk 4]
```diff
@@ -207,11 +207,68 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
...
+
+        sm::make_derive("protocol_errors", _stats.errors[exceptions::exception_code::PROTOCOL_ERROR],
+                        sm::description("Counts the total number of raised CQL PROTOCOL errors.")),
+
+        sm::make_derive("authentication_errors", _stats.errors[exceptions::exception_code::BAD_CREDENTIALS],
+                        sm::description("Counts the total number of raised CQL AUTHENTICATION errors.")),
```
Comment (nyh) [inline transport/server.cc:222]: Nitpick: this is the only case in this patch where the message differs from the name of the error code. I think here too you should write "BAD_CREDENTIALS" in te description, not "AUTHENTICATION". Unless the name BAD_CREDENTIAL itself is a mistake?

File: transport/server.hh (hunk 1 above)
Comment (nyh) [reply transport/server.hh:140]: I agree. It would have been useful if there were a bunch of separate error counters, but if they are just one variable - a map - and it's called "errors", then it's rather obvious.

**Review by nyh — CHANGES_REQUESTED** [review]

File: transport/server.cc [hunk 5]
```diff
@@ -207,11 +207,68 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
         sm::make_derive("requests_shed", _stats.requests_shed,
                         sm::description("Holds an incrementing counter with the requests that were shed due to overload (threshold configured via max_concurrent_requests_per_shard). "
                                             "The first derivative of this value shows how often we shed requests due to overload in the \"CQL transport\" component.")),
-       sm::make_gauge("requests_memory_available", [this] { return _memory_available.current(); },
+        sm::make_gauge("requests_memory_available", [this] { return _memory_available.current(); },
```
Comment (avikivity) [inline transport/server.cc:210]: Why this change?

File: transport/server.cc [hunk 6]
```diff
@@ -207,11 +207,68 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
...
+
+        sm::make_derive("config_errors", _stats.errors[exceptions::exception_code::CONFIG_ERROR],
+                        sm::description("Counts the total number of raised CQL CONFIG errors.")),
+
+        sm::make_derive("already_exists_errors", _stats.errors[exceptions::exception_code::ALREADY_EXISTS],
+                        sm::description("Counts the total number of raised CQL ALREADY_EXISTS errors.")),
+
+        sm::make_derive("unprepared_errors", _stats.errors[exceptions::exception_code::UNPREPARED],
+                        sm::description("Counts the total number of raised CQL UNPREPARED errors.")),
+
```
Comment (avikivity) [inline transport/server.cc:271]: Maybe there should be just one metric, errors, with a label for the exception code.

Adding so many metrics fills me with dread.

File: transport/server.cc [hunk 7]
```diff
@@ -207,11 +207,68 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
...
+        sm::make_derive("unathorized_errors", _stats.errors[exceptions::exception_code::UNAUTHORIZED],
+                        sm::description("Counts the total number of raised CQL UNATHORIZED errors.")),
+
+        sm::make_derive("invalid_errors", _stats.errors[exceptions::exception_code::INVALID],
+                        sm::description("Counts the total number of raised CQL INVALID errors.")),
+
+        sm::make_derive("config_errors", _stats.errors[exceptions::exception_code::CONFIG_ERROR],
+                        sm::description("Counts the total number of raised CQL CONFIG errors.")),
+
+        sm::make_derive("already_exists_errors", _stats.errors[exceptions::exception_code::ALREADY_EXISTS],
```
Comment (avikivity) [inline transport/server.cc:266]: [] can throw. Will probably wreak havoc on the metric collector.

File: transport/server.cc (hunk 6 above)
Comment (amnonh) [reply transport/server.cc:271]: Adding to @avikivity and @nyh it's best to add only one metrics: errors_total (using a suffix that indicates the unit is  good practice)
with a label per error type.
Things to note:
1. It's best to have metrics initialized to zero at an early stage.
2. Labels are mutually exclusive, each error should be counted once under one of the labels, this way summing over all the labels values acts as having one global counter for all the combine errorss.
3. when you add the different metrics with their specific labels, you need to use the same description.

File: transport/server.hh (hunk 1 above)
Comment (StarostaGit) [reply transport/server.hh:140]: Yes, it's a leftover from before the `map` was used, my mistake

File: transport/server.cc (hunk 3 above)
Comment (StarostaGit) [reply transport/server.cc:216]: Not very adamant on the choice of that word. In fact, you might be right that 'returned' is more straightforward in its meaning

File: transport/server.cc (hunk 4 above)
Comment (StarostaGit) [reply transport/server.cc:222]: I used the name of the error code `0x0100` specified in the CQL v4 Protocol (https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v4.spec - section 9). Don't know whether there's a reason our name for it differs.
Maybe somebody knows some more context?

File: transport/server.cc (hunk 5 above)
Comment (StarostaGit) [reply transport/server.cc:210]: Shouldn't it be aligned with the rest of the metrics?
It's such a detrimental change that I included it in this commit since I'm editing the metrics anyway

File: transport/server.cc (hunk 6 above)
Comment (StarostaGit) [reply transport/server.cc:271]: Looking at the labels then

File: transport/server.cc (hunk 7 above)
Comment (StarostaGit) [reply transport/server.cc:266]: It's a small map, how likely are we to actually get something thrown out of there?
I'm now looking at labels to see if that changes what I'm using to keep track of the errors, but if not, what would you suggest I do instead? List each stat as separate member?

File: transport/server.cc (hunk 5 above)
Comment (avikivity) [reply transport/server.cc:210]: We usually make changes minimal.

**Comment by StarostaGit** [comment]
Addressed feedback and put everything into labels

File: transport/server.cc [hunk 8]
```diff
@@ -212,6 +218,101 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
                             seastar::format("Holds the amount of available memory for admitting new requests (max is {}B)."
                                             "Zero value indicates that our bottleneck is memory and more specifically - the memory quota allocated for the \"CQL transport\" component.", _max_request_size))),
 
+        sm::make_derive("cql_errors_total", sm::description("Counts the total number of returned CQL errors."),
+                        error_label_instances.contains(exceptions::exception_code::SERVER_ERROR)
+                        ? std::vector<sm::label_instance>({error_label_instances.at(exceptions::exception_code::SERVER_ERROR)}) : std::vector<sm::label_instance>{},
+                        [this] { return _stats.errors.contains(exceptions::exception_code::SERVER_ERROR) ? _stats.errors.at(exceptions::exception_code::SERVER_ERROR) : 0; }),
```
Comment (amnonh) [inline transport/server.cc:224]: I don't think you need the contains in the map, it will be initialized to zero

Comment (avikivity) [reply transport/server.cc:224]: But the initialization can throw.

Comment (avikivity) [reply transport/server.cc:224]: This is a double lookup, better to use find() and examine the returned iterator.

File: transport/server.cc [hunk 9]
```diff
@@ -159,6 +159,12 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
 {
     namespace sm = seastar::metrics;
 
+    sm::label cql_error_label("type");
+    std::unordered_map<exceptions::exception_code, sm::label_instance> error_label_instances;
```
Comment (amnonh) [inline transport/server.cc:163]: no need for that, just add everything.

File: transport/server.cc [hunk 10]
```diff
@@ -212,6 +218,101 @@ cql_server::cql_server(distributed<cql3::query_processor>& qp, auth::service& au
                             seastar::format("Holds the amount of available memory for admitting new requests (max is {}B)."
                                             "Zero value indicates that our bottleneck is memory and more specifically - the memory quota allocated for the \"CQL transport\" component.", _max_request_size))),
 
+        sm::make_derive("cql_errors_total", sm::description("Counts the total number of returned CQL errors."),
```
Comment (amnonh) [inline transport/server.cc:221]: It's not going to work this way
1. always register the metrics:
a.  you will not get a second chance to do it
b. you can not register the same metrics twice (two metrics with no label)
c. it's better to start a metrics early

File: transport/server.cc (hunk 9 above)
Comment (StarostaGit) [reply transport/server.cc:163]: you mean get rid of the map and put there 19 label instances? It seems rather ugly imo
And also, if anything changes in the `exceptions` namespace, creating the instances like this is just one less place to be affected

File: transport/server.cc (hunk 10 above)
Comment (StarostaGit) [reply transport/server.cc:221]: Alright, I understand when I return the empty label vector it's gonna create problems. And that I should initialize all entries early.
But is there a metric that is not registered? I don't get what you're getting at with 1 and a.

Comment (amnonh) [reply transport/server.cc:221]: in the original code, if a metric was not used yet (i.e. it will be used in the future) you will never register it.
The metrics registration code is called once, if it was not register then, it will never ber register.

I hope at least b,c, convinced you :)

File: transport/server.cc (hunk 9 above)
Comment (amnonh) [reply transport/server.cc:163]: but in your code later you manually call them, so it does not save anything, if your code would have use a loop, I can understand why you do it.

File: transport/server.cc (hunk 8 above)
Comment (StarostaGit) [reply transport/server.cc:224]: I'd say it's a negligible difference when it comes to hashing ints. And this is somewhat more readable. But I can live with changing that.
Although, do we actually need all that stuff if the entries were initialized beforehand? It's true for the `error_label_instances` and can be true for the `_stats.errors` (I just set all of them to 0 in the same place I'm creating the label instances hashmap)

Comment (amnonh) [reply transport/server.cc:224]: @avikivity I probably missing something, if you call `_stats.errors[key]`, wouldn't it return zero if `key` is not in errors?

File: transport/server.cc (hunk 10 above)
Comment (StarostaGit) [reply transport/server.cc:221]: This just makes me more confused. What do you mean by "use" the metric? How can I use it before registering it?
I feel like I'm misunderstanding the terms here - isn't what I'm doing here "registering"? I thought when I call `add_group` and pass a metric created with `sm::make_derive` for example, it's what leads to registration of said metric.

File: transport/server.cc (hunk 9 above)
Comment (StarostaGit) [reply transport/server.cc:163]: I'm more keen on putting creating the metric list for the `cql` group in the loop as well then. Will do that

File: transport/server.cc (hunk 8 above)
Comment (StarostaGit) [reply transport/server.cc:224]: It can resize the table underneath and throw if it can't do it

Comment (StarostaGit) [reply transport/server.cc:224]: At least that's the reason I'm aware of

File: transport/server.cc (hunk 10 above)
Comment (amnonh) [reply transport/server.cc:221]: yes, the registration part (this code) is called once in this case during the contractor.
if you condition your metrics registration, i.g. you only register a metric if it was used, it will never be registered.

let me try to explain it differently
you have a map of errors and your map is empty and an entry is only added on the first time it is used (on the first error) you have a problem with the registration.
If you condition a registration on the existing of an entry, any entry that will not be available during the registration will never be registered even when it will be added at a future time.

Comment (StarostaGit) [reply transport/server.cc:221]: ok, that I understand
but I fail to see how this is conditioned on anything (apart from the situation where I return a separate label instance vector, but you mentioned that in a different point)

Comment (StarostaGit) [reply transport/server.cc:221]: Because in case of the error counting map, we attach the lambda to the metric and only that lambda checks the entry's existence. Is this where I'm mistaken?

Comment (amnonh) [reply transport/server.cc:221]: I gave you three reasons why not to condition your metrics registration on an entry existence, I think we've passed that and we agree you're not going to do it

**Comment by StarostaGit** [comment]
Refactored the code a bit, in response to the above discussions

File: transport/server.cc (hunk 8 above)
Comment (avikivity) [reply transport/server.cc]: Yes

**Comment by avikivity** [comment]
What were the resolutions? Are you always registering the metrics or not?

**Comment by StarostaGit** [comment]
> 
> 
> What were the resolutions? Are you always registering the metrics or not?

Yes, add the metrics in a loop now, so no lookup for the labels is needed. And changed the lambdas to use the found iterator instead of a contains/at pair.
